Desktop_Interface/isobuffer.cpp

```diff
@@ -2,391 +2,364 @@
-                front = bufferEnd;
-                idx = (int) round(((double) front - accumulatedDelay));
-            }
-            readData[i] = buffer[idx];
-
-            accumulatedDelay += timeBetweenSamples;
-        }
-    }
-
-    return readData;
+	/* Refactor Note:
+	 *
+	 * Refactoring this function took a few passes were i made some assumptions:
+	 *  - round() should be replaced by floor() where it was used
+	 *  - int(floor(x)) and int(x) are equivalent (since we are always positive)
+	 *  - free(NULL) is a no-op. This is mandated by the C standard, and virtually all
+	 * implementations comply. A few known exceptions are:
+	 *    - PalmOS
+	 *    - 3BSD
+	 *    - UNIX 7
+	 *   I do not know of any non-compliant somewhat modern implementations.
+	 *
+	 * The expected behavior is to cycle backwards over the buffer with a stride of
+	 * timeBetweenSamples steps, and insert the touched elements into readBuffer.
+	 *
+	 * ~Sebastian Mestre
+	 */
+	const double timeBetweenSamples = sampleWindow * samplesPerSecond / numSamples;
+	const int delaySamples = delayOffset * samplesPerSecond;
+
+	free(readData);
+	readData = (short*) calloc(numSamples, sizeof(short));
+
+	double itr = delaySamples + 1;
+	for (int i = 0; i < numSamples; i++)
+	{
+		while (itr > insertedCount)
+			itr -= insertedCount;
+
+		readData[i] = bufferAt(int(itr));
+
+		if (singleBit)
+		{
+			int subIdx = 8*(-itr-floor(-itr));
+			readData[i] &= (1 << subIdx);
+		}
+
+		itr += timeBetweenSamples;
+	}
+
+	return readData;
 }
 
 void isoBuffer::clearBuffer()
 {
-    for (int i=0; i<bufferEnd;i++){
-        buffer[i] = 0;
-    }
+	for (int i=0; i<bufferEnd; i++)
+	{
+		buffer[i] = 0;
+	}
 
-    back = 0;
-    firstTime = true;
+	back = 0;
 }
 
 void isoBuffer::gainBuffer(int gain_log)
 {
-    qDebug() << "Buffer shifted by" << gain_log;
-    for (int i=0; i<bufferEnd; i++){
-        if (gain_log == -1) buffer[i] *= 2;
-        else buffer[i] /= 2;
-    }
+	qDebug() << "Buffer shifted by" << gain_log;
+	for (int i=0; i<bufferEnd; i++)
+	{
+		if (gain_log == -1) buffer[i] *= 2;
+		else buffer[i] /= 2;
+	}
 }
 
-
+// NOTE: glitch insert is a no-op
 void isoBuffer::glitchInsert(short type)
 {
-
 }
 
-void isoBuffer::enableFileIO(QFile *file, int samplesToAverage, qulonglong max_file_size){
+void isoBuffer::enableFileIO(QFile* file, int samplesToAverage, qulonglong max_file_size)
+{
 
-    //Open the file
-    file->open(QIODevice::WriteOnly);
-    currentFile = file;
+	//Open the file
+	file->open(QIODevice::WriteOnly);
+	currentFile = file;
 
-    //Add the header
-    char headerLine[256];
-    sprintf(headerLine, "EspoTek Labrador DAQ V1.0 Output File\nAveraging = %d\nMode = %d\n", samplesToAverage, virtualParent->driver->deviceMode);
-    currentFile->write(headerLine);
+	//Add the header
+	char headerLine[256];
+	sprintf(headerLine, "EspoTek Labrador DAQ V1.0 Output File\nAveraging = %d\nMode = %d\n", samplesToAverage, virtualParent->driver->deviceMode);
```

Comment:
Putting this in a constant string in an anonymous namespace at the top of the .cpp would be a good idea, as otherwise it's easy to forget to change the string when the algorithm is updated.